integration/tbot_test.go

```diff
@@ -0,0 +1,224 @@
+		TargetPhase: phase,
+	})
+	if err != nil {
+		log.WithError(err).Infof("Error occurred during triggering rotation: %s", phase)
+	}
+	require.NoError(t, err)
+	log.Infof("Triggered rotation: %s", phase)
+
+	if phase == types.RotationPhaseInit {
+		err = waitForProcessEvent(svc(), service.TeleportPhaseChangeEvent, 30*time.Second)
+		require.NoError(t, err)
+	} else {
+		err = waitForProcessEvent(svc(), service.TeleportReloadEvent, 30*time.Second)
+		require.NoError(t, err)
+
+		select {
+		case <-reload:
+		case <-time.After(20 * time.Second):
+			require.FailNow(t, "failed waiting for the service restart")
+		}
+	}
+
+	err = waitForProcessEvent(svc(), service.TeleportReadyEvent, 30*time.Second)
+	require.NoError(t, err)
+
+	err = waitForProcessEvent(svc(), service.ProxyReverseTunnelReady, 30*time.Second)
+	require.NoError(t, err)
+}
+
+func setupServerForCARotationTest(ctx context.Context, t *testing.T, log utils.Logger,
+	errC chan error) (auth.ClientI, func() *service.TeleportProcess, *config.FileConfig, chan struct{},
+) {
+	fc, fds := testhelpers.DefaultConfig(t)
+
+	cfg := service.MakeDefaultConfig()
+	require.NoError(t, config.ApplyFileConfig(fc, cfg))
+	cfg.FileDescriptors = fds
+	cfg.Log = log
+	cfg.CachePolicy.Enabled = false
+	cfg.Proxy.DisableWebInterface = true
+	cfg.RotationConnectionInterval = 500 * time.Millisecond
+	cfg.PollingPeriod = time.Second
+
+	svcC := make(chan *service.TeleportProcess)
+	go func() {
+		err := service.Run(ctx, *cfg, func(cfg *service.Config) (service.Process, error) {
+			svc, err := service.NewTeleport(cfg)
+			if err == nil {
+				svcC <- svc
+			}
+			return svc, err
+		})
+		errC <- err
+	}()
+
+	var svc *service.TeleportProcess
+	select {
+	case <-time.After(30 * time.Second):
+		// this should really happen quite quickly, but under the load during
+		// parallel test run, it can take a while.
+		t.Fatal("teleport process did not instantiate in 30 seconds")
+	case svc = <-svcC:
+	}
+
+	// Ensure the service starts correctly the first time before proceeding
+	_, err := svc.WaitForEventTimeout(30*time.Second, service.TeleportReadyEvent)
+	// in reality, the auth server should start *much* sooner than this.  we use a very large
+	// timeout here because this isn't the kind of problem that this test is meant to catch.
+	require.NoError(t, err, "auth server didn't start after 30s")
+
+	// reloadChan is a blocking channel that releases after each service reload.
+	reloadChan := make(chan struct{})
+	// Tracks the latest instance of the Teleport service through reloads
+	activeSvc := svc
+	activeSvcMu := sync.Mutex{}
+	go func() {
+		for {
+			select {
+			case svc := <-svcC:
+				activeSvcMu.Lock()
+				activeSvc = svc
+				activeSvcMu.Unlock()
+				reloadChan <- struct{}{}
+			case <-ctx.Done():
+				if ctx.Err() != context.Canceled {
+					errC <- ctx.Err()
+				} else {
+					// Push nil to unlock the channel.
+					errC <- nil
+				}
+				return
+			}
+		}
+	}()
```

Comment:
I think this extra channel, goroutine, mutex, and captured variable are not necessary and make things more complicated. `rotate` already wants to assert that reloads happen and be able to get the service before and after the reload without races. You could return svcC (or some type that wraps it) and require `rotate` to explicitly unblock the reloads with a chan receive